### PR TITLE
Add specific tests for {RO}Span.IndexOfAny<char>

### PR DESCRIPTION
**Pull request by benaadams**
Supporting https://github.com/dotnet/coreclr/pull/19790

/cc @krwq

### CODE REVIEW

File: src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+            {
+                index = span.IndexOfAny(searchFor[0], searchFor[1]);
+            }
+            else if (searchFor.Length == 3)
+            {
+                index = span.IndexOfAny(searchFor[0], searchFor[1], searchFor[2]);
+            }
+            else
+            {
+                index = span.IndexOfAny(searchFor);
```
Comment (krwq) [inline src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs]: I think this overload should always be called

Comment (krwq) [reply src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs]: i.e.:
```csharp
            index = span.IndexOfAny(searchFor);

            if (searchFor.Length == 1)
            {
                Assert.Equal(index, span.IndexOf(searchFor[0]));
            }
            else if (searchFor.Length == 2)
            {
                Assert.Equal(index,  span.IndexOfAny(searchFor[0], searchFor[1]));
            }
            else if (searchFor.Length == 3)
            {
               Assert.Equal(index, span.IndexOfAny(searchFor[0], searchFor[1], searchFor[2]));
            }
```

File: src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+        [Fact]
+        public static void ZeroLengthIndexOfTwo_Char()
+        {
+            ReadOnlySpan<char> sp = new ReadOnlySpan<char>(Array.Empty<char>());
+            int idx = sp.IndexOfAny<char>((char)0, (char)0);
+            Assert.Equal(-1, idx);
+        }
+
+        [Fact]
+        public static void DefaultFilledIndexOfTwo_Char()
```
Comment (krwq) [inline src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs]: I'm not sure what is this test testing.

**Comment by benaadams** [comment]
@dotnet-bot test UWP CoreCLR x64 Debug Build

**Comment by benaadams** [comment]
Raised issue for UWP failure https://github.com/dotnet/corefx/issues/32083

File: src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+        [InlineData("aaaaaaaaaaalmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz", "lmr", 'l', 11)]
+        [InlineData("aaaaaaaaaaacmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz", "lmr", 'm', 12)]
+        [InlineData("aaaaaaaaaaarmnopqrstuvwxyzabcdefghijklmnopqrstuvwxyz", "lmr", 'r', 11)]
+        [InlineData("/localhost:5000/PATH/%2FPATH2/ HTTP/1.1", " %?", '%', 21)]
+        [InlineData("/localhost:5000/PATH/%2FPATH2/?key=value HTTP/1.1", " %?", '%', 21)]
+        [InlineData("/localhost:5000/PATH/PATH2/?key=value HTTP/1.1", " %?", '?', 27)]
+        [InlineData("/localhost:5000/PATH/PATH2/ HTTP/1.1", " %?", ' ', 27)]
+        public static void IndexOfAnyStrings_Char(string raw, string search, char expectResult, int expectIndex)
+        {
+            var span = raw.AsSpan();
```
Comment (ahsonkhan) [inline src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs]: nit: remove use of var

File: src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+            char[] searchFor = search.ToCharArray();
+
+            var index = span.IndexOfAny(searchFor);
```
Comment (ahsonkhan) [inline src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs]: same here, nit: remove use of var

File: src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs (hunk 3 above)
Comment (benaadams) [reply src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs]: Updated

File: src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs (hunk 4 above)
Comment (benaadams) [reply src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs]: Updated

**Comment by benaadams** [comment]
@danmosemsft you triggered a rebuild with your comment, looks like infra not happy
```
10:36:10 Done building Native components
10:36:20 -- The C compiler identification is MSVC 19.14.26433.0
10:36:22 -- The CXX compiler identification is MSVC 19.14.26433.0
10:36:22 -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.14.26428/bin/Hostx86/x86/cl.exe
10:36:23 Cannot contact windows.10.amd64.clientrs4.devex.open-0913102446322-1: hudson.remoting.RequestAbortedException: java.nio.channels.ClosedChannelException
12:24:57 Cancelling nested steps due to timeout
12:24:57 Could not connect to windows.10.amd64.clientrs4.devex.open-0913102446322-1 to send interrupt signal to process
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // timeout
[Pipeline] }
[Pipeline] // timestamps
[Pipeline] echo
Archiving logs in netci-archived-logs/**
[Pipeline] step
Required context class hudson.Launcher is missing
Perhaps you forgot to surround the code with a step that provides this, such as: node
[Pipeline] echo
Error during cleanup: java.lang.IllegalArgumentException: Failed to prepare archiveArtifacts step
[Pipeline] isUnix
Required context class hudson.Launcher is missing
Perhaps you forgot to surround the code with a step that provides this, such as: node
[Pipeline] echo
Some files could not be cleaned up because of running processes.  These processes will be killed immediately and cleanup will happen before the node upon node-reuse
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
[BFA] Scanning build for known causes...
[BFA] No failure causes found
[BFA] Done. 0s
Setting status of 658a5716713374ceaed027b4970b4876a04c125f to FAILURE with url https://ci3.dot.net/job/dotnet_corefx/job/master/job/windows-TGroup_netcoreapp+CGroup_Release+AGroup_x86+TestOuter_false_prtest/16424/ and message: 'Build finished. '
Using context: Windows x86 Release Build
Timeout has been exceeded
Finished: ABORTED
```

**Comment by benaadams** [comment]
@dotnet-bot test NETFX x86 Release Build 
@dotnet-bot test Windows x86 Release Build

**Comment by danmosemsft** [comment]
@mmitche why do random comments sometimes trigger rebuilds?

**Comment by mmitche** [comment]
@danmosemsft The comment triggers a jenkins webhook, which causes it to compare the last built sha against the current sha.  That may not be up to date, and it may cause a rebuild.  Jenkins also forgets that info over time (as builds are deleted, restarts, etc.) which makes a rebuild for an old PR more likely.

File: src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Memory/tests/ReadOnlySpan/IndexOfAny.char.cs]: Yep, can't make sense of it anymore. I will have to dig up what the test used to be.

I think the intention was to look for (default, x) and (x, default) in a span containing only default, and assert we always find it at the first index.

Therefore, the inner loop is unnecessary, at least the way the test is written atm:
`for (int i = 0; i < length; i++)`

It should just be single check. The test should look like this:
```C#
            for (int length = 0; length < byte.MaxValue; length++)
            {
                byte[] a = new byte[length];
                Span<byte> span = new Span<byte>(a);

                 int idx = span.IndexOfAny(default, 99);
                 Assert.Equal(0, idx);

                 idx = span.IndexOfAny(99, default);
                 Assert.Equal(0, idx);
            }
```

We can leave it as is for now and fix all the tests (byte/char/etc.) in one go outside of this PR.

**Comment by ahsonkhan** [comment]
> is this good to merge now?

I haven't gotten a chance to fully review all the tests yet, but does this PR need to wait for https://github.com/dotnet/coreclr/pull/19790 to get merged first?

**Comment by benaadams** [comment]
>  does this PR need to wait for dotnet/coreclr#19790 to get merged first?

No; its just a copying of the `byte` tests to `char` to help validate dotnet/coreclr#19790; so before would probably be better :)

**Comment by ViktorHofer** [comment]
@dotnet-bot test this please

File: src/System.Memory/tests/ReadOnlySpan/IndexOfAny.byte.cs [hunk 5]
```diff
@@ -36,22 +36,18 @@ public static void IndexOfAnyStrings_Byte(string raw, string search, char expect
             char[] searchFor = search.ToCharArray();
             byte[] searchForBytes = Encoding.UTF8.GetBytes(searchFor);
 
-            var index = -1;
+            var index = span.IndexOfAny(new ReadOnlySpan<byte>(searchForBytes));
```
Comment (stephentoub) [inline src/System.Memory/tests/ReadOnlySpan/IndexOfAny.byte.cs]: Nit: var => int (I realize it was like that before... but it shouldn't have been :)

File: src/System.Memory/tests/ReadOnlySpan/IndexOfAny.byte.cs [hunk 6]
```diff
@@ -32,11 +32,11 @@ public static partial class ReadOnlySpanTests
         public static void IndexOfAnyStrings_Byte(string raw, string search, char expectResult, int expectIndex)
         {
             byte[] buffers = Encoding.UTF8.GetBytes(raw);
-            var span = new ReadOnlySpan<byte>(buffers);
+            ReadOnlySpan<byte> span = new ReadOnlySpan<byte>(buffers);
```
Comment (stephentoub) [inline src/System.Memory/tests/ReadOnlySpan/IndexOfAny.byte.cs]: FWIW, this was fine as `var`.  Our guidelines just outlaw it when the type isn't named on the RHS, which generally means something other than a new or a cast.  But naming it in such situations is fine, too.

**Comment by stephentoub** [comment]
@ahsonkhan, anything more to do here?

**Comment by benaadams** [comment]
UWP infra issue
```
NuGet.targets(114,5): error : Failed to download package 'System.Linq.4.1.0' from 'https://api.nuget.org/v3-flatcontainer/system.linq/4.1.0/system.linq.4.1.0.nupkg'. [D:\j\workspace\windows-TGrou---c60886e1\external\test-runtime\XUnit.Runtime.depproj] [D:\j\workspace\windows-TGrou---c60886e1\external\test-runtime\XUnit.Runtime.depproj]
NuGet.targets(114,5): error : Response status code does not indicate success: 504 (Gateway Timeout). [D:\j\workspace\windows-TGrou---c60886e1\external\test-runtime\XUnit.Runtime.depproj] [D:\j\workspace\windows-TGrou---c60886e1\external\test-runtime\XUnit.Runtime.depproj]
```

**Comment by benaadams** [comment]
@dotnet-bot test UWP NETNative x86 Release Build

**Pull request merged**